File: CppEIM/src/distributions/GMM.cpp
#include <distributions/GMM.h>

GMM::GMM() :
    weight_distribution(vec(1, fill::ones)){
}

GMM::GMM(vec weights, mat means, cube covars) :
    weight_distribution(weights),
    dim(means.n_rows),
    num_components(weights.n_rows){
    mat covar_0 = covars.slice(0);
    for (uword i = 0; i < num_components; ++i){
        components.push_back(Gaussian(means.col(i), covars.slice(i)));
    }
}

vec GMM::density(const mat &samples) const {
    vec densities(samples.n_cols, fill::zeros);
    vec p = weight_distribution.probabilities();
    for (uword i = 0; i < num_components; ++i){
        densities += p.at(i) * components[i].density(samples);
    }
    return densities;
}

vec GMM::log_density(const mat &samples) const {
    return log(density(samples) + 1e-25);
}

double GMM::log_likelihood(const mat &samples) const{
    return mean(log_density(samples));
}

mat GMM::sample(uword size) const{
    uvec cts = weight_distribution.sample_cts(size);
    mat samples = mat(dim, size, fill::none);

    uvec end_idx = cumsum(cts);
    uvec start_idx = end_idx - cts;

    #pragma omp parallel for default(none) shared(cts, samples, start_idx, end_idx)
    for(uword i = 0; i < num_components; ++i){
        if (cts(i) > 0) {
            samples.cols(start_idx.at(i), end_idx.at(i) - 1) = components[i].sample(cts(i));
        }
    }
    uvec perm = shuffle(linspace<uvec>(0, size - 1, size));
    return samples.cols(perm);
}

File: EIM/distributions/marginal/GMM.py
import numpy as np
from distributions.marginal.Categorical import Categorical
from distributions.marginal.Gaussian import Gaussian


class GMM:

    def __init__(self, weights, means, covars):
        self._weight_distribution = Categorical(weights)
        self._components = [Gaussian(means[i], covars[i]) for i in range(means.shape[0])]

    def density(self, samples):
        densities = np.stack([self._components[i].density(samples) for i in range(self.num_components)], axis=0)
        w = np.expand_dims(self.weight_distribution.probabilities, axis=-1)
        return np.sum(w * densities, axis=0)

    def log_density(self, samples):
        return np.log(self.density(samples) + 1e-25)

    def log_likelihood(self, samples):
        return np.mean(self.log_density(samples))

    def sample(self, num_samples):
        w_samples = self._weight_distribution.sample(num_samples)
        samples =[]
        for i in range(self.num_components):
            cns = np.count_nonzero(w_samples == i)
            if cns > 0:
                samples.append(self.components[i].sample(cns))
        return np.random.permutation(np.concatenate(samples, axis=0))

    @property
    def components(self):
        return self._components

    @property
    def num_components(self):
        return len(self._components)

    @property
    def weight_distribution(self):
        return self._weight_distribution
